OAK-10294 - Add to indexing job the Pipelined Strategy to dump Mongo contents

nfsantos opened the pull request:
Add a Pipelined strategy to create the FlatFileStore from MongoDB during the indexing job.

nfsantos committed:
Initial commit of pipelined download strategy.

nfsantos committed:
Minor fixes

nfsantos committed:
WIP

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into GRANITE-45911

nfsantos committed:
Add recovery from broken MongoDB connections to downloader.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into GRANITE-45911

nfsantos committed:
Shutdown gracefully if one of the dump stages fails with an exception.

nfsantos committed:
Add documentation.
Change the configuration for the retrial mechanism for MongoDB connection failures: instead of number of retrials, use the amount of time to keep trying before giving up.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Add support for auto-tuning working set memory based on total available memory to the indexer.
Switch to using system properties to configure the pipelined strategy (instead of env variables).

nfsantos committed:
Always use read preference secondaryPreferred for downloading from Mongo.

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedStrategy.java:
```diff
@@ -0,0 +1,403 @@
+
+    private final Comparator<PipelinedNodeStateHolder> comparator;
+    private final PathElementComparatorStringArray pathComparator;
+    private NotificationEmitter emitter;
```
this is never used. Are you planning to use it in the future?

nfsantos replied:
It was copied from the old traverse strategy. It is part of the mechanism to monitor free memory using the management MBeans, which the Pipelined strategy does not use. I removed this field.

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+        try {
+            LOG.info("Starting to download from MongoDB.");
+            MongoCollection<BasicDBObject> dbCollection = MongoDocumentStoreHelper.getDBCollection(mongoStore, collection);
+            //TODO This may lead to reads being routed to secondary depending on MongoURI
```
I wonder if this might result in problems if the secondary is far behind

nfsantos replied:
I'm not sure. The MongoDocumentStore implementation has logic to limit the reads from secondaries if they are too far back. Maybe we can implement a similar logic here. But I don't know what could be the problems, maybe it will miss some of the recent documents?

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+                        // Reset the retries
+                        failureTimestamp = null;
+                        // Continue downloading everything starting from the next _lastmodified value
+                        downloadRange(new TraversingRange(new LastModifiedRange(nextLastModified + 1, upperBoundLastModified), null));
```
I think there is a small probability that nextLastModified + 1 is larger than upperBoundLastModified. In this case, the constructor of LastModifiedRange might throw an exception:

>         if (lastModifiedTo < lastModifiedFrom) {
>             throw new IllegalArgumentException("Invalid range (" + lastModifiedFrom + ", " + lastModifiedTo + ")");
>         }

nfsantos replied:
I think it cannot happen. `upperBoundLastModified` is set to 1 plus the highest _modified value that can be returned by the traversal queries to MongoDB:
```
long upperBoundLastModified = getLatestLastModified(dbCollection) + 1;
```
And `nextLastModified` is only assigned to values returned by the query:
```
this.nextLastModified = next.getLong(NodeDocument.MODIFIED_IN_SECS);
```
Furthermore, the traversing query is exclusive of the upper bound, see `TraversingRange.getFindQuery()`, so it will always return `_modfied` values lower than `upperBoundLastModfied`.

So at most, `nextLastModifed+1` will be equal to `upperBoundLastModified`. Or is there something wrong with my reasoning?

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+                }
+            }
+        } catch (Throwable t) {
+            LOG.warn("Thread terminating with exception: " + t);
```
I would log the exception as well: ", t)"

nfsantos replied:
I'm trying to avoid polluting the logs with multiple repetitions of the same stack trace. The exception here is propagated and caught by the main thread that is waiting for these tasks to complete (In the PipelinedStrategy class). And then the main thread rethrows the exception wrapped in a RuntimeException. The code that called the `createdSortedStoreFile()` will catch and finally log the exception with the stack trace.

Actually, there is a corner case where the exception will not be logged completely, which is if the main thread already stopped waiting for some of these futures, for instance, if one of the tasks aborted which will cause the main thread to abort with the exception of the first task that failed. In this case, we would only get the name and message of the exceptions thrown by the futures that failed later. Ok, that stack trace may be useful, so I will log the stack trace here as well (except for InterruptedExceptions).

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+                .limit(1)
+                .iterator()) {
+            if (!cursor.hasNext()) {
+                return -1;
```
It's an edge case... but what about adding a parameter "defaultValue" so that -1 is never used? Let's say in the first call it returns -1, and then a document is added... Very strange edge case indeed.

nfsantos replied:
This should never happen, it would mean the collection of nodes is empty. What about throwing an exception in this case? In any case, if the collection is empty, there is nothing to index, so we can as well fail the job.

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedNodeStateHolderFactory.java:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * Factory of {@link PipelinedNodeStateHolder}, reuse an internal array list to reduce object allocation.
+ * Not thread safe.
```
As a general comment (not asking to change it):

I wonder if what is more important: (A) thread save + simple code, or (B) add this complexity, risk problems later because it is not thread save, and avoid memory allocation, hopefully speeding it up a bit? 

In this case I would go for simpler and thread-safe code, and only try to optimize if we see it shows up in profiling output. Maybe you have done that already. There are cases where optimizations can't be done later easily, but here it can be. See also https://lemire.me/blog/2023/04/27/hotspot-performance-engineering-fails/

For speed, an alternative is to call PathUtils.getDepth(path) first to calculate the array size, and then construct an array of the right size, and then fill the entries one by one? Sure, we would iterate over the array twice.

nfsantos replied:
This code is on the hot path, so I am willing to sacrifice simplicity for efficiency. However, I have not run any benchmark to validate that creating a new intermediate array for every line read from the intermediate files will cause a measurable performance degradation. Intuitively, I expect so, because this is a tight loop and the files contain millions of lines, but I could be wrong.

I did not knew about the `PathUtils.getDepth(path)`, that seems to be a better solution as it avoids the need for any intermediate object and can be done in a thread-safe way. I will re-rewrite the code to use that method to create an array of the right size, as you suggest, therefore making it thread safe. If I have time, run some microbenchmarks.

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PathElementComparatorStringArray.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Comparator;
+import java.util.Set;
+
+class PathElementComparatorStringArray implements Comparator<String[]> {
```
I think we are missing a unit test here.

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+        this.collection = collection;
+        this.batchSize = batchSize;
+        // TODO
+//        this.filter = filter;
```
This TODO should be resolved

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+
+        //TODO This may lead to reads being routed to secondary depending on MongoURI
+        //So caller must ensure that its safe to read from secondary
+//        this.readPreference = MongoDocumentStoreHelper.getConfiguredReadPreference(mongoStore, collection);
```
This TODO should be resolved

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedSortBatchTask.java:
```diff
@@ -0,0 +1,151 @@
+                emptyBuffersQueue.put(nseBuffer);
+            }
+        } catch (Throwable t) {
+            LOG.warn("Thread terminating with exception: " + t);
```
Log the exception ", t"

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedTransformTask.java:
```diff
@@ -0,0 +1,215 @@
+                }
+            }
+        } catch (Throwable t) {
+            LOG.warn("Thread terminating with exception: " + t);
```
Log the exception

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/SortKeyPathFactory.java:
```diff
@@ -0,0 +1,65 @@
+    private final ArrayList<String> arrayBuilder = new ArrayList<>(16);
+    // Common words that appear in paths.
+    // TODO: confirm that checking for the common words is a worthwhile optimization.
+    private static final Set<String> commonWords = Set.of("content", "dam", "product-assets",
```
We are missing a unit test

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/SortKeyPathFactory.java:
```diff
@@ -0,0 +1,65 @@
+//            "videoCodec", "audioCodec", "metadata", "original", "profile"
+    );
+
+    public String[] genSortKey(String path) {
```
generateSortKey?

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/SortKeyPathFactory.java:
```diff
@@ -0,0 +1,65 @@
+            // It is not worth to intern all levels because at lower levels the names are more likely to be less diverse,
+            // often even unique, so interning them would fill up the interned string hashtable with useless entries.
+            if (i < 3 || part.length() == 1 || commonWords.contains(part)) {
+                arrayBuilder.add(part.intern());
```
Sure we can save a bit of memory, but I think it is not worth the complexity.

I think part.length() is not very common.

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PathElementComparatorStringArray.java:
```diff
@@ -0,0 +1,70 @@
+                return 1;
+            }
+
+            String pe1 = p1[i1]; i1++;
```
I would use String pe2 = p2[i2++]; 
as it is less prone to typos

thomasmueller commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java:
```diff
@@ -0,0 +1,278 @@
+    private final MongoCollection<BasicDBObject> dbCollection;
+    private final ReadPreference readPreference;
+
+    private long documentsRead = 0;
```
Initializing to 0 / null is not needed. And it is inconsistent.

fabriziofortino reviewed (approved):
LGTM.
The new strategy is a lot less complicated than the previous ones. The code is a lot easier to understand. It would be great to add some tests.

nfsantos committed:
Address reviews comments.

nfsantos committed:
Use 0 and Long.MAX_VALUE as boundaries for download range, no need to query Mongo to determine the earliest and latest values of _modified.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Add a new stage to merge sorted files.
Log how much time the download and transform threads spend waiting to enqueue their outputs in the out queues (this indicates that the stages after them are too slow).
Collect and log metrics describing the download and transform stages, like mongo documents downloaded, node state entries extracted, filtered, mongo documents that do not match a node state (garbage) and a few others.
Use only FileUtils.byteCountToDisplaySize() to pretty print values in bytes, as the humanReadableByteCount was using 1000 for 1K instead of 1024.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Improve collection and logging of metrics.

nfsantos committed:
Add unit test for merge-sort task.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Add more tests, including a first draft of an integration test.
Refactor code.

nfsantos committed:
Fix

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Add an integration test

nfsantos committed:
Add license header.

nfsantos committed:
Add license header.

nfsantos committed:
Rename test files.

nfsantos committed:
Do not filter on _modified != null when doing non-recoverable download.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

nfsantos committed:
Fixes based on review comments.

nfsantos committed:
Bound the size of the histograms.

nfsantos committed:
Add missing license files.

nfsantos committed:
Improve logging and code clean of bounded histogram.

nfsantos committed:
Display human readable byte counts with 2 decimal places instead of abbreviating to closest integer value.

nfsantos committed:
Improved logging.

nfsantos committed:
Refactoring

nfsantos committed:
Remove changes unrelated to PR.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10294

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/NodeStateHolder.java:
```diff
@@ -0,0 +1,54 @@
+    public NodeStateHolder(String line) {
+        this.line = line;
+        String path = getPath(line);
+        int depth = PathUtils.getDepth(path);
```
minor: for consistency you might want to statically import `getDepth`.

nfsantos replied:
I have instead replaced all static imports by references of the form `<class>.<method>`. I find this more readable.

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMergeSortTask.java:
```diff
@@ -0,0 +1,140 @@
+import static org.apache.jackrabbit.oak.index.indexer.document.flatfile.pipelined.PipelinedStrategy.SENTINEL_SORTED_FILES_QUEUE;
+
+/**
+ * Accumulates the intermediate sorted files and, when all files are genereted, merges them into a single sorted file,
```
typo
```suggestion
 * Accumulates the intermediate sorted files and, when all files are generated, merges them into a single sorted file,
```

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMergeSortTask.java:
```diff
@@ -0,0 +1,140 @@
+        Stopwatch w = Stopwatch.createStarted();
+        File sortedFile = new File(storeDir, getSortedStoreFileName(algorithm));
+        try (BufferedWriter writer = createWriter(sortedFile, algorithm)) {
+            Function<String, NodeStateHolder> func1 = (line) -> line == null ? null : new NodeStateHolder(line);
+            Function<NodeStateHolder, String> func2 = holder -> holder == null ? null : holder.getLine();
```
can you rename these with more meaningful names?

nfsantos replied:
Done, I use the same names as the arguments defined in the function that is called with these variables: `stringToType` and `typeToString`.

fabriziofortino commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java:
```diff
@@ -0,0 +1,304 @@
+    /**
+     * Whether to retry on connection errors to MongoDB.
+     * This property affects the query that is used to download the documents from MongoDB. If set to true, the query
+     * will traverse the results by order of the _modfied property (does an index scan), which allows it to resume after
```
typo
```suggestion
     * will traverse the results by order of the _modified property (does an index scan), which allows it to resume after
```

fabriziofortino commented on oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/NodeStateEntryBatchTest.java:
```diff
@@ -0,0 +1,97 @@
+
+public class NodeStateEntryBatchTest {
+
+    @Test
+    public void testMaximumNumberOfEntries() {
+        NodeStateEntryBatch batch = NodeStateEntryBatch.createNodeStateEntryBatch(1024, 2);
+        assertFalse(batch.isAtMaxEntries());
+        batch.addEntry("a", new byte[1]);
+        assertFalse(batch.isAtMaxEntries());
+        batch.addEntry("b", new byte[1]);
+        assertEquals(2, batch.numberOfEntries());
+        assertTrue(batch.isAtMaxEntries());
+        try {
+            batch.addEntry("c", new byte[1]);
+            throw new AssertionError("Expected exception");
+        } catch (IllegalStateException e) {
+            // expected
+        }
+    }
```
what about using `expected`? Same for the test below.
```suggestion
    @Test(expected = IllegalStateException.class)
    public void testMaximumNumberOfEntries() {
        NodeStateEntryBatch batch = NodeStateEntryBatch.createNodeStateEntryBatch(1024, 2);
        assertFalse(batch.isAtMaxEntries());
        batch.addEntry("a", new byte[1]);
        assertFalse(batch.isAtMaxEntries());
        batch.addEntry("b", new byte[1]);
        assertEquals(2, batch.numberOfEntries());
        assertTrue(batch.isAtMaxEntries());
        batch.addEntry("c", new byte[1]);
    }```

nfsantos replied:
I replaced by something like:
```assertThrows(BufferOverflowException.class, () -> batch.addEntry("b", new byte[1]));```
which is supported since JUnit 4.13. Oak is using 4.13.1.

steffenvan commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedStrategy.java:
```diff
@@ -0,0 +1,424 @@
+            int maxNumberOfEntriesPerBuffer = maxNumberOfEntries / numberOfBuffers;
+
+            // A ByteBuffer can be at most Integer.MAX_VALUE bytes
+            int bufferSizeBytes;
```
Just for readability purposes, could the assignment here with the local scope be refactored into a function?

nfsantos replied:
Done

fabriziofortino commented on oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedIT.java:
```diff
@@ -0,0 +1,149 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class PipelinedIT {
```
can we have a test where the `pathPredicate` does not match any node in the repository? In the past, this has caused issues because the files where empty.

nfsantos replied:
Done

fabriziofortino reviewed (approved):
LGT. Just a few improvements.

nfsantos committed:
Add tests for error handling when configuration properties have invalid values.
Minor refactoring.

nfsantos committed:
Update oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMergeSortTask.java

Co-authored-by: Fabrizio Fortino <fabrizio.fortino@gmail.com>

nfsantos committed:
Update oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java

Co-authored-by: Fabrizio Fortino <fabrizio.fortino@gmail.com>

nfsantos committed:
Add test for when the path predicate does not match any existing path in the document store.
Address other review comments.

nfsantos committed:
Merge remote-tracking branch 'origin/OAK-10294' into OAK-10294

nfsantos committed:
Add unit test to test recovery from broken mongo connections.
Fix: when a connection to Mongo is lost, the documents that were collected in a block but not yet enqueued were being lost.

steffenvan commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedStrategy.java:
```diff
@@ -0,0 +1,433 @@
+                LOG.info("Waiting for tasks to complete.");
+                int tasksFinished = 0;
+                int transformTasksFinished = 0;
+                while (tasksFinished < numberOfThreads) {
```
Would it be possible to refactor some of the code in this function into more separate functions? The function itself is quite long and with 3 nested `try-catch blocks`. While this might be necessary, I think it would at least improve the readability for the future. I asked ChatGPT to do it for an older version of the code, and the result was quite nice. 

For example: 
```java
private void waitForTasksToComplete(int numberOfThreads, int transformThreads, int entryCount, Stopwatch start,
        ArrayBlockingQueue<BasicDBObject[]> mongoDocQueue,
        ArrayBlockingQueue<NodeStateEntryBatch> emptyBatchesQueue,
        ArrayBlockingQueue<NodeStateEntryBatch> nonEmptyBatchesQueue,
        ArrayBlockingQueue<File> sortedFilesQueue,
        TransformStageStatistics transformStageStatistics) {
        int tasksFinished = 0;
        int transformTasksFinished = 0;
        while (tasksFinished < numberOfThreads) {
            Future<?> completedTask = ecs.take();
         ...
```

And then we could call it:

```java
try {
    waitForTasksToComplete(ecs, numberOfThreads, monitorFuture, mongoDocQueue, emptyBatchesQueue,
        nonEmptyBatchesQueue, sortedFilesQueue, transformStageStatistics, start);
    flatFileStore = mergeSortedFilesTask.getFlatFileStoreFile();
    return flatFileStore;
} finally {
    shutdownExecutors();
}
```

nfsantos replied:
I agree that the method is very long and somewhat hard to read. But the refactoring proposed above is also not very satisfying, as the method has too many arguments. I'll keep this in mind for a future revision of this class, hopefully we can come up with a cleaner and more logical refactor. 

For the final changes to this PR, I'd rather focus only on correctness or on small changes, and avoid doing any refactoring other than trivial ones.

nfsantos committed:
Add missing license.
